siatest/testgroup.go

```diff
@@ -0,0 +1,365 @@
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewGroup creates a group of TestNodes from node params. All the nodes will
+// be connected, synced and funded. Hosts nodes are also announced.
+func NewGroup(nodeParams ...node.NodeParams) (*TestGroup, error) {
+	// Create and init group
+	tg := &TestGroup{
+		nodes:   make(map[*TestNode]struct{}),
+		hosts:   make(map[*TestNode]struct{}),
+		renters: make(map[*TestNode]struct{}),
+		miners:  make(map[*TestNode]struct{}),
+	}
+
+	// Create node and add it to the correct groups
+	nodes := make([]*TestNode, 0, len(nodeParams))
+	for _, np := range nodeParams {
+		node, err := NewCleanNode(np)
+		if err != nil {
+			return nil, err
+		}
+		// Add node to nodes
+		tg.nodes[node] = struct{}{}
+		nodes = append(nodes, node)
+		// Add node to hosts
+		if np.Host != nil || np.CreateHost {
+			tg.hosts[node] = struct{}{}
+		}
+		// Add node to renters
+		if np.Renter != nil || np.CreateRenter {
+			tg.renters[node] = struct{}{}
+		}
+		// Add node to miners
+		if np.Miner != nil || np.CreateMiner {
+			tg.miners[node] = struct{}{}
+		}
+	}
+
+	// Fully connect nodes
+	if err := fullyConnectNodes(nodes); err != nil {
+		return nil, err
+	}
+
+	// Get a miner and mine some blocks to generate coins
+	if len(tg.miners) == 0 {
+		return nil, errors.New("cannot fund group without miners")
+	}
+	miner := tg.Miners()[0]
+	for i := types.BlockHeight(0); i <= types.MaturityDelay; i++ {
+		if err := miner.MineBlock(); err != nil {
+			return nil, err
+		}
+	}
+	// Fund nodes
+	if err := fundNodes(miner, tg.nodes); err != nil {
+		return nil, err
+	}
+	// Add storage to hosts
+	if err := addStorageFolderToHosts(tg.hosts); err != nil {
+		return nil, err
+	}
+	// Announce hosts
+	if err := announceHosts(tg.hosts); err != nil {
+		return nil, err
+	}
+	// Mine a block to get the announcements confirmed
+	if err := miner.MineBlock(); err != nil {
+		return nil, err
+	}
+	// Block until all hosts show up as active in the renters' hostdbs
+	if err := hostsInRenterDBCheck(miner, tg.renters, len(tg.hosts)); err != nil {
+		return nil, err
+	}
+	// Set renter allowances
+	if err := setRenterAllowances(tg.renters); err != nil {
+		return nil, err
+	}
+	// Wait for all the renters to form contracts
+	if err := waitForContracts(miner, tg.renters, tg.hosts); err != nil {
+		return nil, err
+	}
+	// Make sure all nodes are synced
+	if err := synchronizationCheck(miner, tg.nodes); err != nil {
+		return nil, err
+	}
+	return tg, nil
+}
+
+// setRenterAllowances sets the allowance of each renter
+func setRenterAllowances(renters map[*TestNode]struct{}) error {
+	for renter := range renters {
+		if err := renter.RenterPost(defaultAllowance); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// NewGroupFromTemplate will create hosts, renters and miners according to the
+// settings in groupParams.
+func NewGroupFromTemplate(groupParams GroupParams) (*TestGroup, error) {
```

Comment:
We try to cluster the exported functions (even that don't act on an object) together the same way that we try to cluster functions of an object together.

And by extension, we cluster exported functions of an object together, and the unexported functions of an object are clustered together in a separate cluster.